FontSizePicker: Rewrite unit tests to use userEvent and be more comprehensive

noisysocks opened the pull request:
## What?
Rewrites the `FontSizePicker` unit tests to use `userEvent` where possible and be more comprehensive.

## Why?
I want to make @ciampo happy 🙂 see https://github.com/WordPress/gutenberg/pull/44891#pullrequestreview-1142800395.

## How?
I looked at the `FontSizePicker` component storybook, listed all of the functionality that I could think of, and then wrote new tests from scratch that covered everything that I listed.

## Testing Instructions
`npm run test`, obviously, but the more important thing is to review the code and check that I haven't missed anything.

## Screenshots or screencast <!-- if applicable -->

noisysocks committed:
Rewrite FontSizePicker unit tests to use userEvent and be more comprehensive

ciampo commented on packages/components/src/font-size-picker/test/index.tsx:
```diff
@@ -1,290 +1,434 @@
+			const input = screen.getByLabelText( 'Custom' );
+			await user.clear( input );
+			await user.type( input, '80' );
+			expect( onChange ).toHaveBeenCalledWith( expectedValue );
```
Do we care about checking how many times `onChange` has been called?

(here and for other tests)

noisysocks replied:
It seems to be unpredictable (perhaps it's called once per keystroke and there's some debouncing going on?), or at the very least: I'm too dumb to work out how many times `onChange` is supposed to be called 😅

At any rate I don't think there's much value in testing this. If `onChange` is _eventually_ called with the correct value then I'm happy.

ciampo replied:
From what I understand:

- `clear()` should cause 0 or 1 calls to `onChange`, depending on whether we expect to call `onChange` when `value` changes to empty string (I can't remember off the top of my head)
- `type(80)` should cause another 2 calls to `onChange` — one per keystroke (first `8`, then `0`)

> It seems to be unpredictable 

This is exactly why I'd check the number of times `onChange` fires — in the past, we had regressions on `InputControl` & co. caused by changes in this behaviour that were interfering with the propagation of changes.

> At any rate I don't think there's much value in testing this. If onChange is eventually called with the correct value then I'm happy.

Anyway, while I think there is value in testing it, I also agree that this is a secondary test — if it's too complicated to write a good test for it right now, we can skip it for now, I wouldn't consider it a blocker

ciampo commented on packages/components/src/font-size-picker/test/index.tsx:
```diff
@@ -1,290 +1,434 @@
-			expect( unitSelect ).toBeTruthy();
-			expect( unitLabel ).toBeFalsy();
-			expect( fontSize ).toBe( '16px' );
+	it( 'should not use units when the initial value has no unit', async () => {
```
Isn't this testing for the same behavior as the `should call onChange( $expectedValue ) after user types 80 when value is $value` test (second pair of value/expectedValue) ?

noisysocks replied:
Oops, yes forgot to delete this.

ramonjd commented on packages/components/src/font-size-picker/test/index.tsx:
```diff
@@ -1,290 +1,434 @@
-			it( 'should fallback to font size `slug` for header hint label if `name` is undefined', () => {
-				const fontSizes = [
-					{
-						slug: 'gigantosaurus',
```
Do you have something against gigantosauruses?  🦖

noisysocks replied:
You're right, I'll put her back.

ramonjd commented on packages/components/src/font-size-picker/test/index.tsx:
```diff
@@ -1,290 +1,434 @@
+						value={ value }
+					/>
+				);
+				// TODO: There's no accessible way to select the label.
```
There's been a label since #45180 (slightly modified in #45298) as far as I'm aware.

noisysocks replied:
How did I miss that! I'll query for the existence of an element with that label 👍

ciampo replied:
I was also going to point out that in case you need to do exact test matching across multiple DOM elements, you can pass a custom matcher to `screen.getByText`:

```tsx
function textContentMatcher( textMatch: string | RegExp ) {
	const hasText =
		typeof textMatch === 'string'
			? ( node: Element | null ) => node?.textContent === textMatch
			: ( node: Element | null ) =>
					textMatch.test( node?.textContent ?? '' );

	const matcherFn: MatcherFunction = ( _content, node ) => {
		if ( ! hasText( node ) ) {
			return false;
		}

		const childrenDontHaveText = Array.from( node?.children || [] ).every(
			( child ) => ! hasText( child )
		);

		return childrenDontHaveText;
	};

	return matcherFn;
}

// later, in a test:
const label = screen.getByText(
	textContentMatcher( expectedHint )
);
```

ramonjd commented on packages/components/src/font-size-picker/test/index.tsx:
```diff
@@ -1,290 +1,434 @@
+				const label = document.querySelector(
+					'.components-base-control__label'
+				);
+				expect( label ).toHaveTextContent( expectedHint );
```
The `expectedHint` in this test case doesn't look right to me. 

<img width="928" alt="Screen Shot 2022-10-28 at 12 20 17 pm" src="https://user-images.githubusercontent.com/6458278/198437472-0ea3c1b6-3b89-4686-abba-a5e6344cd436.png">

Could it be that the test is passing because [toHaveTextContent](https://github.com/testing-library/jest-dom#tohavetextcontent) does partial matches?

If we're only interested in testing whether the accessible label and text hint are set correctly, then another option might be to just pass a single fontSize for the label you want to test:

```js
		test.each( [
			{
				value: undefined,
				expectedHint: 'Size Default',
				fontSizePresets: [ fontSizes[ 0 ] ],
			},
			{
				value: '8px',
				expectedHint: 'Size Tiny(px)',
				fontSizePresets: [ fontSizes[ 0 ] ],
			},
			{
				value: '3px',
				expectedHint: 'Size (Custom)',
				fontSizePresets: [ fontSizes[ 0 ] ],
			},
		] )(
			'displays $expectedHint as header hint when value is $value',
			( { value, expectedHint, fontSizePresets } ) => {
				render(
					<FontSizePicker
						__nextHasNoMarginBottom
						fontSizes={ fontSizePresets }
						value={ value }
					/>
				);
				const label = screen.getByLabelText( expectedHint );
				expect( label ).toBeInTheDocument();
				expect( label ).toHaveTextContent(
					// Removes whitespaces to match text interspersed in HTML elements.
					expectedHint.replace( /\s*/g, '' )
				);
			}
		);
```

This avoids have to fire an event to click on preset buttons and what not.

noisysocks replied:
By fixing your other comment I think I inadvertently fixed this, good catch 😀

ramonjd reviewed (approved):
Thanks for improving the unit tests. Good coverage 👍 

I left some questions about testing the header hint and labels.

noisysocks committed:
Use header label's aria-label

noisysocks committed:
Put back the Gigantosaurus 🦖

noisysocks committed:
Update changelog

noisysocks committed:
Merge branch 'trunk' into update/font-size-picker-unit-tests

noisysocks commented:
Thanks @ramonjd! Updated this per your feedback and added a changelog entry.

ciampo commented:
About to post my review as, sorry for the delay!

ciampo commented on packages/components/src/font-size-picker/test/index.tsx:
```diff
@@ -1,290 +1,430 @@
+			);
+			expect( screen.getByLabelText( 'Custom' ) ).toBeInTheDocument();
+			// TODO: onChange() shouldn't be called.
+			//expect( onChange ).not.toHaveBeenCalled();
```
Good find! Adding a comment here so that we don't miss this as a follow-up

noisysocks replied:
It's one of the things we fixed in https://github.com/WordPress/gutenberg/pull/44891 :)

ciampo commented on packages/components/src/font-size-picker/test/index.tsx:
```diff
@@ -1,290 +1,430 @@
+
+		test.each( [
+			{ value: undefined, expectedLabel: 'Size Default' },
+			{ value: '12px', expectedLabel: 'Size small(px)' },
+			{ value: '40px', expectedLabel: 'Size Gigantosaurus(px)' },
```
On a similar note as https://github.com/WordPress/gutenberg/pull/45298#discussion_r1007512570, is this label format the expected one? Seeing "px" without a numeric value almost makes the label look broken (as if the value couldn't be computed correctly)

noisysocks replied:
No, definitely not. https://github.com/WordPress/gutenberg/pull/44966 changes the hint text to match what's in the designs and will remove this. But let's get tests for the existing (incorrect) behaviour in first.

ciampo commented on packages/components/src/font-size-picker/test/index.tsx:
```diff
@@ -1,290 +1,430 @@
+			await user.click(
+				screen.getByRole( 'button', { name: 'Set custom size' } )
+			);
+			// TODO: There are two spinbuttons with similar names, so can't
+			// select the slider's input in an accessible way.
```
Isn't there only 1 spinbutton (the UnitControl to the side of the slider) ? The slider itself should have role="slider" if I remember correctly?

(This applies also to other tests performing the same query)

noisysocks replied:
They don't have `role="slider"` (does such a role exist?) but it is labelled. Not sure why I thought it wasn't. Fixed!

ciampo commented on packages/components/src/font-size-picker/test/index.tsx:
```diff
@@ -1,290 +1,430 @@
+			const sliderInput = document.querySelector(
+				'.components-range-control__slider'
+			);
+			// Fake the slider drag using fireEvent. Testing drag-and-drop using
+			// userEvent is a job for the RangeControl tests.
+			fireEvent.change( sliderInput!, {
+				target: { value: 80 },
+			} );
```
Unfortunately React Testing Library doesn't offer a way to test with drag events yet, so it's probably OK do leave this as-is.

I'd disagree on the "is a job for RangeControl tests" though — if there was a better way to test slider inputs, I'd argue that it should be used across all of our tests (and not just `RangeControl`). This type of consistency is great and it allows the "good patterns" to spread and be adopted around the codebase!

noisysocks replied:
Fair enough! I'll remove my editorialising comment 😛

ciampo reviewed (commented):
Thank you @noisysocks for working on this!!

Apart from the inline comments, it looks like we're not adding tests for:

- "complex" CSS values (e.g `clamp(1.75rem, 3vw, 2.25rem)`)
- font sizes with different units (i.e some in `px`, some in `rem`, ...)

Those are probably some of the harder to test edge cases, and therefore it would be nice to have unit tests for those too!

> I want to make @ciampo happy

I'm flattered 😊

noisysocks committed:
Remove editorialising comment

noisysocks committed:
Use getByLabelText( 'Custom Size' ) to select slider

noisysocks committed:
Remove redundant test

noisysocks commented:
Addressed some of the feedback. I'll add more tests for non-px values tomorrow.

noisysocks committed:
Add tests for non-px font sizes

noisysocks committed:
Assert length of options/radios

noisysocks committed:
Assert number of calls to onChange

noisysocks committed:
Merge branch 'trunk' into update/font-size-picker-unit-tests

noisysocks commented:
Okey dokey, I've added tests for non-`px` and `clamp()` values and added assertions for the number of times `onChange` is called. Should be good to go 👍

noisysocks commented:
Merging this so I can unblock https://github.com/WordPress/gutenberg/pull/44891 and https://github.com/WordPress/gutenberg/pull/44966. Thanks for the feedback @ciampo!

ciampo commented:
Thank YOU for always taking the time to make significant improvements to the components you work on!
